src/librustc/ty/subst.rs

```diff
@@ -314,7 +314,37 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
 
 impl<'tcx> TypeFoldable<'tcx> for &'tcx Substs<'tcx> {
     fn super_fold_with<'gcx: 'tcx, F: TypeFolder<'gcx, 'tcx>>(&self, folder: &mut F) -> Self {
-        let params = self.params.iter().map(|k| k.fold_with(folder)).collect();
+        // The most general way to implement this function is as follows.
+        //
+        //   let params = self.params.iter().map(|k| k.fold_with(folder)).collect();
+        //   folder.tcx().mk_substs(Substs {
+        //       params: params
+        //   })
+        //
+        // But this function is sufficiently hot that it's a sizeable win to
+        // handle some cases specially.
+        // - If folding doesn't change the substs, we avoid calling `mk_substs`
+        //   and instead reuse the existing substs.
+        // - In that case, when the substs length is one, we also avoid
+        //   creating a new Vec, which avoids a heap allocation.
+        let params;
+        if self.params.len() == 1 {
+            let param = self.params[0].fold_with(folder);
+            if param == self.params[0] {
+                return &self;
+            } else {
+                params = vec![param]
+            }
+        } else if self.params.len() > 1 {
+            params = self.params.iter().map(|k| k.fold_with(folder)).collect();
+            if params == self.params {
+                return &self;
+            }
+        } else {
+            return &self;
```

Comment:
super minor, but the `&` in `&self` is redundant. The `Self` type is `&'tcx Substs<'tcx>` and the `self` identifier is of type `&Self`. Deref coercion means that using `&self`, `self` or `*self` here is going to be equivalent.